includes/class-llms-media-protector.php

```diff
@@ -0,0 +1,640 @@
+ * to the upload directory, normally `WP_CONTENT_DIR . '/uploads'`, is saved as '_wp_attached_file' metadata.
+ *
+ * Example of uploading a file:
+ *
+ *     $media = new LLMS_Media_Protector();
+ *     $media->set_upload_subdirectory( $media->get_upload_subdirectory() . '/social-learning' );
+ *     $post_data['meta_input'] = array( LLMS_Media_Protector::AUTHORIZATION_FILTER_KEY => 'llms_sl_authorize_media_view' );
+ *     $id = $media->handle_upload( 'image', 0, $post_data );
+ *
+ * Example of protecting a file:
+ *
+ *     add_filter( 'llms_sl_authorize_media_view', array( $this, 'authorize_media_view' ), 10, 3 );
+ *
+ *     public function authorize_media_view( $is_authorized, $media_id, $url ) {
+ *         $is_authorized = current_user_can( 'view_others_students' );
+ *         return $is_authorized;
+ *     }
+ *
+ * @since [version]
+ *
+ * @todo Add handling of HTTP range requests. See {@see https://datatracker.ietf.org/doc/html/rfc7233} and
+ *       {@see https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests}.
+ */
+class LLMS_Media_Protector {
+
+	/**
+	 * The meta key used to specify the filter hook name that authorizes viewing of a media file.
+	 *
+	 * @TODO Should the key be prefixed with an underscore '_' to denote private?
+	 *
+	 * @since [version]
+	 *
+	 * @var string
+	 */
+	public const AUTHORIZATION_FILTER_KEY = 'llms_media_authorization_filter';
+
+	/**
+	 * The name of the query parameter for whether the media image should be treated as an icon.
+	 *
+	 * @since [version]
+	 *
+	 * @var string
+	 */
+	public const QUERY_PARAMETER_ICON = 'llms_media_icon';
+
+	/**
+	 * The name of the query parameter for the media post ID.
+	 *
+	 * @since [version]
+	 *
+	 * @var string
+	 */
+	public const QUERY_PARAMETER_ID = 'llms_media_id';
+
+	/**
+	 * The name of the query parameter for the requested media image size.
+	 *
+	 * @since [version]
+	 *
+	 * @var string
+	 */
+	public const QUERY_PARAMETER_SIZE = 'llms_media_image_size';
+
+	/**
+	 * Serve the media file by reading and outputting it with the readfile() function.
+	 *
+	 * @since [version]
+	 *
+	 * @var int
+	 */
+	public const SERVE_READ_FILE = 1;
+
+	/**
+	 * Serve the media file by redirecting the HTTP client with a "Location" header.
+	 *
+	 * @since [version]
+	 *
+	 * @var int
+	 */
+	public const SERVE_REDIRECT = 2;
+
+	/**
+	 * Serve the media file by sending an "X-Sendfile" style header and let the HTTP server serve the file.
+	 *
+	 * @since [version]
+	 *
+	 * @var int
+	 */
+	public const SERVE_SEND_FILE = 3;
+
+	/**
+	 * The root path component for uploaded LifterLMS files with a leading slash '/' and without a trailing slash.
+	 *
+	 * @since [version]
+	 *
+	 * @var string
+	 */
+	protected $upload_subdirectory = '/llms-uploads';
+
+	/**
+	 * Set up this class.
+	 *
+	 * @since [version]
+	 *
+	 * @return void
+	 */
+	public function __construct() {
```

Comment:
I'm thinking that it might make sense to add parameters to the constructor to configure the upload directory.

The core defaults of `/llms-uploads` makes sense and I think that our add-ons (and probably 3rd parties) will want to put their own directory inside of that directory. But still guarding against a need to create an entirely separate directory makes sense.

I'm thinking that having a `$base_upload_subdirectory = '/llms-uploads'` and an additional `$upload_subdirectory = ''` would work. The final directory will be `$base_upload_subdirectory . $upload_subdirectory`. The constructor might look like this:

```suggestion
	public function __construct( string $upload_subdirectory = '', string $base_upload_subdirectory = '/llms-uploads' ) {
		// Clean input and set to protected class vars...
```

This would make it so that the example in the docs above could be shortened from:
```
 $media = new LLMS_Media_Protector();
 $media->set_upload_subdirectory( $media->get_upload_subdirectory() . '/social-learning' );
```

To:

```
 $media = new LLMS_Media_Protector(); // Core plugin.
 $media = new LLMS_Media_Protector( '/social-learning' ); // add-on.
 $media = new LLMS_Media_Protector( '', 'custom-dir' ); // Custom directory.
```

I suppose the input could also be an array so that in the last example we don't have to pass an empty string.

Thoughts?

Reply:
@thomasplevy, 

I like the idea of optionally setting the upload directory in the constructor.

Currently, the class is instantiated by `LifterLMS->init()` and the constructor sets up hooks. Perhaps the hooks should be setup in a new method that `LifterLMS->init()` calls. Then the constructor could be used to set the upload directory. Although it looks like setting up hooks multiple times wouldn't hurt (other than minor waste of time).

Another option would be to split the class into one that handles uploads and another that handles authorization, URLs and file serving.

Options:
1. Add setting of the upload directory to the constructor that also sets up hooks.
2. Add setting of the upload directory to the constructor and add a new method that sets up hooks.
3. Split the class into two.

Reply:
It might make sense to split into two different classes, especially if someone wants to perform an upload later they'll need to create a new instance (which sets up the hooks again).

Although it might be better to keep it as it is now but just expose the instance via a class variable. Then `LLMS_Assets()` class works similarly to how I think the media class could work. In this scenario we could probably leave it more or less how it is now without much change and establish a pattern that add-ons could create their own instance within the plugin, setup on init, and then access itself later... Maybe this is what you were going for and then my only concern would be not requiring a new instance every time.

Reply:
I'm leaning away from splitting the class up into a upload class and a download class because most (92%) of the code is for downloading. There are only 3 small methods used to set the authorization filter (per your suggestion above) and change the upload directory.

I had not considered multiple files being uploaded in one request. If different add-ons need to upload at the same time (is this likely?), this would currently require multiple instances of the media protector class and some minor time wasted re-registering hooks. I've implemented option 2 to solve this problem. Option 4 would be to use a singleton pattern.